codegen/attribute_analyzer.go

```diff
@@ -0,0 +1,191 @@
+package codegen
+
+import (
+	"goa.design/goa/expr"
+)
+
+type (
+	// AttributeAnalyzer analyzes an attribute and its sub-attributes.
+	AttributeAnalyzer interface {
+		// IsPointer returns true if the attribute being analyzed is a pointer.
+		IsPointer() bool
+		// IsRequired returns true if the Required property is set to true.
+		IsRequired() bool
+		// DefaultValue returns the default value for the attribute being analyzed.
+		DefaultValue() interface{}
+		// Name returns the type name for the attribute. If withPkg is true, it
+		// prefixes the type name with the package name.
+		Name(withPkg bool) string
+		// Ref returns the type reference for the attribute. If withPkg is true, it
+		// prefixes the type ref with the package name.
+		Ref(withPkg bool) string
+		// Def returns the type definition for the attribute.
+		Def() string
+		// Identifier returns a valid identifier for the name and attribute
+		// expression.
+		Identifier(name string, firstUpper bool) string
+		// Attribute returns the attribute being analyzed.
+		Attribute() *expr.AttributeExpr
+		// Properties returns the attribute properties used in analysis.
+		Properties() *AttributeProperties
+		// Dup creates a shallow copy of the analyzer with the given attribute
+		// and its required property.
+		Dup(att *expr.AttributeExpr, required bool) AttributeAnalyzer
+		// SetAttribute sets the attribute to analyze.
+		SetAttribute(att *expr.AttributeExpr)
+		// SetProperties sets the attribute properties in the analyzer.
+		SetProperties(required, pointer, defPtr, useDefault bool)
+	}
+
+	// AttributeProperties contains properties that affect how the attribute
+	// is stored (i.e. pointer or non-pointer).
+	AttributeProperties struct {
+		// Pointer if true indicates that the attribute is a pointer even if
```

Comment:
Pointer indicates whether the attribute is a pointer ...

Same with comments below.